Pegasus/include/Math.hpp

```diff
@@ -30,13 +659,450 @@ class JacobiEigenvalue
     glm::dmat3 m_eigenvectors;
     glm::dvec3 m_eigenvalues;
 
+    /**
+     * @brief Finds maximal off diagonal matrix element.
+     * @param mat matrix to search
+     * @param i max element row index
+     * @param j max element columnt index
+     */
     static void FindMaxNormOffDiagonal(glm::dmat3 const& mat, uint8_t& i, uint8_t& j);
+
+    /**
+     * @brief Calculates rotation angle for a given matrix and it's element.
+     * @param mat matrix to rotate
+     * @param i element row index
+     * @param j element column index
+     * @return angle in radians
+     */
     double CalculateRotationAngle(glm::dmat3 const& mat, uint8_t i, uint8_t j) const;
+
+    /**
+     * @brief Makes Givens rotation matrix from the angle and indices.
+     * @param theta rotation angle in radians
+     * @param i row index
+     * @param j column index
+     * @return Givens rotation matrix
+     */
     static glm::dmat3 MakeGivensRotationMatrix(double theta, uint8_t i, uint8_t j);
+
+    /**
+     * @brief Calculates eigenvalues and eigenvectors.
+     */
     void Calculate();
 };
 
+/**
+ * @brief Quickhull convex hull calculation algorithm.
+ * @tparam Vertices STL compatible random access glm::dvec3 container
+ */
+template <typename Vertices>
+class QuickhullConvexHull
+{
+public:
+    class Face;
+    using Faces = std::list<Face>;
+
+    /**
+     * @brief Convex hull face container.
+     */
+    class Face
+    {
+    public:
+        /**
+         * @brief Constructs a face without initializing above vertices.
+         * @param hedsFaceIterator HEDS face iterator
+         * @param hyperPlane HyperPlane
+         * @param indices face indices
+         */
+        Face(HalfEdgeDataStructure::face_iterator hedsFaceIterator,
+            HyperPlane const& hyperPlane,
+            std::array<size_t, 3> const& indices
+        )
+            : m_extremalVertexIndex()
+            , m_hedsFaceIterator(hedsFaceIterator)
+            , m_hyperPlane(hyperPlane)
+            , m_indices(indices)
+        {
+            std::sort(m_indices.begin(), m_indices.end());
+        }
+
+        /**
+         * @brief Constructs face and initializes it's abobe vertices.
+         * @param vertexBuffer common vertex buffer
+         * @param partitionMarkedVertices vertices marked for partitioning
+         * @param hedsFaceIterator HEDS face iterator
+         * @param hyperPlane HypePlane
+         * @param indices face indices
+         */
+        Face(Vertices& vertexBuffer,
+            std::list<typename Vertices::iterator>& partitionMarkedVertices,
+            HalfEdgeDataStructure::face_iterator hedsFaceIterator,
+            HyperPlane const& hyperPlane,
+            std::array<size_t, 3> const& indices
+        )
+            : Face(hedsFaceIterator, hyperPlane, indices)
+        {
+            SetVertices(vertexBuffer, partitionMarkedVertices);
+        }
+
+        /**
+         * @brief Partitions vertices that are above the face.
```

Comment:
Describe what happens to buffer and list (since they are references I assume they are changed inside the method)